tests/test_blurb_add.py

```diff
@@ -0,0 +1,147 @@
+from itertools import chain, product
+import re
+
+import pytest
+
+from blurb import blurb
+
+
+ALLOWED_ISSUE_URL_PREFIX = [
+    'github.com/python/cpython/issues/',
+    'http://github.com/python/cpython/issues/',
+    'https://github.com/python/cpython/issues/'
+]
+
+ALLOWED_SECTION_IDS = list(map(str, range(1 + len(blurb.sections), 1)))
+
+
+def test_valid_no_issue_number():
+    assert blurb._extract_issue_number(None) is None
+    res = blurb._update_blurb_template(issue=None, section=None)
+    lines = res.splitlines()
+    assert f'.. gh-issue:' not in lines
+    assert f'.. gh-issue: ' in lines
+    for line in lines:
+        assert not line.startswith('.. section: ')
+
+
+@pytest.mark.parametrize(('issue', 'expect'), [
+    (f'{w1}{prefix}12345{w2}', '12345')
+    for (w1, w2) in product(['', ' '], repeat=2)
+    for prefix in ('', 'gh-', *ALLOWED_ISSUE_URL_PREFIX)
+])
```

Comment:
I'm not keen on these super-condensed parameterisations, tests are "special" and often benefit from duplication and being more explicit:

> **Since tests don't have tests, it should be easy for humans to manually inspect them for correctness,** even at the expense of greater code duplication. This means that the DRY principle often isn’t a good fit for unit tests, even though it is a best practice for production code.
>
> **In tests we can use the [DAMP principle](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests) (“Descriptive and Meaningful Phrases”), which emphasizes _readability over uniqueness_.** Applying this principle can introduce code redundancy (e.g., by repeating similar code), but it makes tests more obviously correct.

https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

https://stackoverflow.com/q/129693/724176

Side-by-side input and corresponding expected results are much clearer than having to decode `itertools.product` and unroll two loops; or `itertools.chain` and `zip` and five loops below.

Reply:
There was indeed a mistake where I replaced underscores by whitespaces instead of whitespaces by underscores ! In general, I like using `chains` instead of `+` but I agree that `product` might be a bit to harsh. For checking identity-like section names, I still kept the `zip` but it's a single `zip` without anything else.